src/cluster.c

```diff
@@ -4224,7 +4232,7 @@ void clusterBeforeSleep(void) {
      * including replicas. */
     if (moved_slot >= 0) {
```

Comment:
No we discussed how it should be sent and if it should be in a separate channel. (We can't send "MOVED -2 ip:port". If I just remove the if, the code will crash.)

Reply:
> If I just remove the if, the code will crash

Curious - was it Redis that crashed or the test? Any challenge/risk in fixing this crash?

> it should be in a separate channel

That sounds a bit counter-intuitive to me, if I as a client need to subscribe to two different channels to get the same type of information. The difference between the two channels sounds more like an implementation detail than something conceptual. I was under the impression that the separate channel would be used for different types of events, such as the arrival/departure of a replica.

Reply:
> > If I just remove the if, the code will crash
> 
> Curious - was it Redis that crashed or the test? Any challenge/risk in fixing this crash?

clusterNotifyMovedSlot accesses `server.cluster->slots[moved_slot]`. With a negative index, it's read outside the allocation. Without the  `if >= 0`, it'd also try with -1 when there is no slot moved. I must assume you misunderstand. It needs more implementation if we want to notify for multiple slots and in particular, it needs to be decided what the notification should look like.

> > it should be in a separate channel
> 
> That sounds a bit counter-intuitive to me, if I as a client need to subscribe to two different channels to get the same type of information. The difference between the two channels sounds more like an implementation detail than something conceptual. I was under the impression that the separate channel would be used for different types of events, such as the arrival/departure of a replica.

I thought you preferred a separate channel in this comment: https://github.com/redis/redis/pull/10358#discussion_r818893065

Reply:
So, exactly what behaviour do you suggest for multiple moved slots? This often means failover but the cluster bus doesn't specify that.

Reply:
I misunderstood your [comment](https://github.com/redis/redis/pull/10358#discussion_r818426335). I thought it was meant for different types of events. 

I think there should be one channel for each type of messages and the type discussed here is "slot ownership". The reason why the slot ownership changes isn't that important to the client (be it slot migration or failover) IMO. The use case I have in mind is to facilitate faster/cheaper routing information update.

Reply:
OK, now we sorted out the misunderstandings. :-)

In the case of failover, we might have something like 5000 slots moved at once. Should we send 5000 messages to the subscribers in that case? It is flooding the clients with too much data if we do it in the naive way. We need to send a more compact message which includes a list of slot ranges.

.... or we can just send a message like "RELOAD" (or "INVALIDATE" or "RESHARDED") which tells the client to reload the slot mapping using CLUSTER SLOTS, SHARDS or NODES, whatever the client prefers.

... or we can send no message at all, which will result in a -MOVED redirect sooner or later, which triggers the client to reload the slot mapping. This is my initial choice in this PR, because it is minimalistic and extensible.